tools/src/main/scala/scala/scalanative/linker/Link.scala

```diff
@@ -3,14 +3,33 @@ package linker
 
 import scalanative.nir._
 import scalanative.util.Scope
+import scala.scalanative.io.VirtualDirectory
+import java.nio.file.Path
 
 object Link {
 
+  private val cache =
+    collection.mutable.HashMap[(Path, Seq[Global], Seq[Long]), Result]()
+
   /** Load all clases and methods reachable from the entry points. */
   def apply(config: build.Config, entries: Seq[Global])(implicit
       scope: Scope
-  ): Result =
-    Reach(config, entries, ClassLoader.fromDisk(config))
+  ): Result = {
+    val classpath = config.classPath.map { path =>
+      ClassPath(VirtualDirectory.real(path))
+    }
+    val classPathMtime = classpath.map(_.lastModifiedMillis)
+    val key = (config.buildPath, entries, classPathMtime)
```

Comment:
The current implementation would leak the memory (each  result would probably add tens of MBs which would never be released). Java WeakHashMap should help with that to some extent (in fact SoftReference based solution would work much better here, but there seems to HashMap implementation based on that in the stdlib). 
Alternatively, the cache should be cleaned every time we know that it could never be reused again. I think that we might keep the cache when changing the main class, but the classpath is the same. In any other case, we should do a clean build. 

If the classpath has not changed and the main class is the same, we would always get exactly the same list entries `Seq[Global]` equal to mainClass and fixed amount of CodeGen/Interflow/Lower dependencies. The only thing that might have changed is the main class, and it already covered  by the `config.buildPath`. Similary if we produce a dynamic library instead of application, then if the classpath has not changed, then the list of entries would not change as well.
Also as soon as at least 1 file of classpath was modified, the outputs of the program might have changed, and we need to rebuild it. Because of that we can keep just the most recent modification time for all the classpath files.

Based on that the key could be simplified to `(buildPath: Path, lastModifiedTime: Long)`

Reply:
In general we should be very fragile when it comes to increasing memory usage. The problems with OOMs while build Scala Native projects is already a known issue in some of OSS  projects.

Reply:
> In general we should be very fragile when it comes to increasing memory usage.

Well, it's worth distinguishing between local development (where caching is very valuable for incremental re-building and rapid development) and CI (where we only build once and memory limits are very strict). Scala.js linker config has a flag to distinguish these situations, which we configure in our build plugins.

https://github.com/scala-js/scala-js/blob/5323e68c2daeb2937f807f0fd2559cbd3e1b6e71/linker-interface/shared/src/main/scala/org/scalajs/linker/interface/StandardConfig.scala#L55-L66

Reply:
> The current implementation would leak the memory (each result would probably add tens of MBs which would never be released). Java WeakHashMap should help with that to some extent (in fact SoftReference based solution would work much better here, but there seems to HashMap implementation based on that in the stdlib).

Oh, yes, that's correct.
I attempted to use `WeakHashMap`, but it appears that the cache entries can be easily garbage collected because no strong references to the keys remain (It would be more ideal if there was a SoftReference-based HashMap in the standard library, as you pointed out...).

As an alternative, perhaps we don't need to store multiple cache entries, but just retain the most recent one, especially considering the incremental compilation scenario.

Like in the form of `private var cache: (Key, Result) = ...`.?



> Based on that the key could be simplified to (buildPath: Path, lastModifiedTime: Long)

Thanks! Indeed, we don't need to use `entries` as we already use the timestamps of classPaths + buildPath 👍 

---

> In general we should be very fragile when it comes to increasing memory usage. The problems with OOMs while build Scala Native projects is already a known issue in some of OSS projects.

That makes sense.
Hmm, we can reduce the number of cache entries, decrease the object size of keys, and make the cache easier to garbage collect. However, the in-memory cache still increases memory usage by storing the `Linker.Result` (I'm not sure how much it could increase though) for some extent.

It would be better to check how much the `Linker.Result` eats up the memory if we store in a cache, or should we store the Result in file so we don't store them in-memory?

<details>
<summary>I attempted to check the cache size using jol but somehow I couldn't use it 🤔 </summary>

```diff
diff --git a/project/Deps.scala b/project/Deps.scala
index 69a0f761..9d75f86e 100644
--- a/project/Deps.scala
+++ b/project/Deps.scala
@@ -20,6 +20,7 @@ object Deps {
   lazy val ScalaCheck          = "org.scalacheck"         %% "scalacheck"                 % "1.15.4"
   lazy val ScalaTest           = "org.scalatest"          %% "scalatest"                  % "3.2.9"
   lazy val ScalaParCollections = "org.scala-lang.modules" %% "scala-parallel-collections" % "1.0.3"
+  lazy val Jol                 = "org.openjdk.jol"         % "jol-core"                   % "0.17"
   lazy val SbtPlatformDeps     = "org.portable-scala"      % "sbt-platform-deps"          % "1.0.1"
   lazy val SbtTestInterface    = "org.scala-sbt"           % "test-interface"             % "1.0"
   lazy val JUnitInterface      = "com.novocode"            % "junit-interface"            % "0.11"
@@ -28,8 +29,10 @@ object Deps {
   def Tools(scalaVersion: String) = {
     List(ScalaCheck % "test", ScalaTest % "test") ++
       scalaVersionsDependendent(scalaVersion) {
@@ -28,8 +29,10 @@ object Deps {
   def Tools(scalaVersion: String) = {
     List(ScalaCheck % "test", ScalaTest % "test") ++
       scalaVersionsDependendent(scalaVersion) {
-        case (2, 12) => Nil
-        case _       => ScalaParCollections :: Nil
+        case (2, 12) => Jol :: Nil
+        case _       => ScalaParCollections :: Jol :: Nil
       }
   }
   def NativeLib(scalaVersion: String) = scalaVersionsDependendent(scalaVersion) {
```

```
[info] compiling 1 Scala source to /Users/tanishiking/ghq/github.com/tanishiking/scala-native/project/target/scala-2.12/sbt-1.0/classes ...
[error] /Users/tanishiking/ghq/github.com/tanishiking/scala-native/tools/src/main/scala/scala/scalanative/linker/Link.scala:8:20: object jol is not a member of package org.openjdk
[error] import org.openjdk.jol.info.GraphLayout;
[error]                    ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
[warn] Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? (default: r)
```

</details>

---

> Scala.js linker config has a flag to distinguish these situations, which we configure in our build plugins.

I'm not sure how much things we store in memory for incremental compilation in scala-native at this moment, but it seems good to have for scala-native too if we store a lot things in memory for inc-compilation.

Reply:
To fix problem with jol you would need to add that dependency to `project/build.sbt` That's because to start sbt project we need to build sbt-plugin and it's dependencies first. So, to break that cyclic dependency we bootstrap them when starting the sbt. The only downside is the fact that we need to duplicate some dependencies config.

Reply:
Thanks! I printed debug using `jol` and it turned out the build cache has around 35MB (with 1 cache entry) even for the `sandbox` project IIUC. Which seems not a good idea to keep in memory.

Now I'm wondering which is the way to go for build cache

- Cache the NIR linking/optimization in memory (which takes more than 30-40MB, it may go more for the real projects)
- Dump the `LInker.Result` into a disk for cache
  - I'm not sure this is really a good idea, reading / writing that size of stuff may take some CPU and memory.
- Or, maybe we should (if cache hits) skip the whole output of NIR linking ~ CodeGen

<details>

```diff
diff --git a/project/build.sbt b/project/build.sbt
index f161c052..27ba3856 100644
--- a/project/build.sbt
+++ b/project/build.sbt
@@ -18,6 +18,7 @@ addSbtPlugin("com.github.sbt" % "sbt-pgp" % "2.2.0")
 addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.4.4")

 libraryDependencies += "org.eclipse.jgit" % "org.eclipse.jgit.pgm" % "5.10.0.202012080955-r"
+libraryDependencies += "org.openjdk.jol" % "jol-core" % "0.17"

 // scalacOptions used to bootstrap to sbt prompt.
 // In particular, no "-Xfatal-warnings"
diff --git a/tools/src/main/scala/scala/scalanative/linker/Link.scala b/tools/src/main/scala/scala/scalanative/linker/Link.scala
index a8acf077..a2338442 100644
--- a/tools/src/main/scala/scala/scalanative/linker/Link.scala
+++ b/tools/src/main/scala/scala/scalanative/linker/Link.scala
@@ -5,6 +5,8 @@ import scalanative.nir._
 import scalanative.util.Scope
 import scala.scalanative.io.VirtualDirectory
 import java.nio.file.Path
+import org.openjdk.jol.info.GraphLayout
+

 object Link {

@@ -15,6 +17,7 @@ object Link {
   def apply(config: build.Config, entries: Seq[Global])(implicit
       scope: Scope
   ): Result = {
+    println(GraphLayout.parseInstance(cache).totalSize())
     val classpath = config.classPath.map { path =>
       ClassPath(VirtualDirectory.real(path))
     }
```

```
sbt:scala-native> sandbox2_13/nativeLink
# WARNING: Unable to get Instrumentation. Dynamic Attach failed. You may add this JAR as -javaagent manually, or supply -Djdk.attach.allowAttachSelf
# WARNING: Unable to attach Serviceability Agent. Unable to attach even with module exceptions: [org.openjdk.jol.vm.sa.SASupportException: Sense failed., org.openjdk.jol.vm.sa.SASupportException: Sense failed., org.openjdk.jol.vm.sa.SASupportException: Sense failed.]
120
[info] Linking (multithreading false) (867 ms)
[info] Checking intermediate code (79 ms)
[info] Dumping intermediate code (linked) (212 ms)
[info] Discovered 778 classes and 4879 methods
[info] Optimizing (debug mode) (594 ms)
[info] Checking intermediate code (37 ms)
[info] Dumping intermediate code (optimized) (140 ms)
[info] Dumping intermediate code (lowered) (230 ms)
[info] Generating intermediate code (587 ms)
[info] Produced 39 files
[info] Compiling to native code (189 ms)
[info] Linking native code (immix gc, none lto) (3 ms)
[info] Total (2745 ms)
[success] Total time: 4 s, completed May 17, 2023, 6:18:15 PM
sbt:scala-native> sandbox2_13/nativeLink
36837256
[info] NIR linking skipped due to cache hit.
[info] Linking (multithreading false) (1055 ms)
[info] Checking intermediate code (24 ms)
[info] Dumping intermediate code (linked) (127 ms)
[info] Discovered 778 classes and 4879 methods
[info] Optimizing (debug mode) (466 ms)
[info] Checking intermediate code (22 ms)
[info] Dumping intermediate code (optimized) (215 ms)
[info] Dumping intermediate code (lowered) (195 ms)
[info] Generating intermediate code (398 ms)
[info] Produced 39 files
[info] Compiling to native code (178 ms)
[info] Linking native code (immix gc, none lto) (3 ms)
[info] Total (2496 ms)
[success] Total time: 3 s, completed May 17, 2023, 6:18:34 PM
```

</details>

Reply:
I think we start with going with the Scala.js approach - they do not cache the results of linking, but they read IR instead. That might make more sense, because any change of inputs, would lead to modification of inputs. However, by keeping the IR in memory they reduce the costs of IO and reading from the disk every time. 
I'm in the progress of rebasing Denys's changes to compact the NIR file size. With this in mind we might use one of the two approaches in the future: keep the deserialized Defns in-memory (scala.js seems to be doing that) or keep the binary NIR in-memory and deserialize it on demand (no IO overhead, lower memory usage, but would require additional cycles for deserialization).

Scala.js implementation is a bit convoluted, but that might be a good starting point: the cache is based on `IRContainers` (jars or directories) and results in `IRFile`s mapping to a single `ir.Tree`. In our case that would be a `Seq[nir.Defn]` or `Map[nir.Global, nir.Defn]` because we keep the flat structure at that point.
https://github.com/scala-js/scala-js/blob/5323e68c2daeb2937f807f0fd2559cbd3e1b6e71/linker-interface/shared/src/main/scala/org/scalajs/linker/interface/IRFileCache.scala#L23

Despite the previous comments, I think we might have 3 levels of caching: 
- config + classpath based - allowing to skip the whole build process if nothing has changed, just return previously linked binary
- NIR cache - to skip unnecessary IO for parts of the classpath that have not changed
- Compilation outputs cache - (we already have that when building with experimental `NativeConfig.incrementalCompilation)), it skips a generation of LLVM IR for parts of the code that have not changed (currently package scoped)

Everything between linking (Reach) and emitting LLVM IR cannot be safely cached unless we would invest heavily in implementing incremental optimizer, and incremental code lowering (involves reimplementation of virtual method dispatch). Otherwise, all these steps are very fragile to any change of the inputs. The next step after that might be incremental linking, but that might be harder to achieve.

Reply:
>config + classpath based - allowing to skip the whole build process if nothing has changed, just return previously linked binary

We have the additional complication that we allow native code in jars and sub-projects that is not present in Scala.js. 

- If the classpath itself does not change and no Scala files in the sub-projects change then no classes or NIR are created then the front end of the build can be skipped as appropriate. This is handled by the build system and is before the compile itself starts.
- We still need to run the backend as native code can change. We currently hash the jar and directories and store the result but there is some work involved to check the hash. We only unpack the code and copy to the build area the first time or if the hash differs. Perhaps at that point we could short circuit the rest of the build rather than calculating time stamps which does take some time. I think that would have to be passed along from phase to phase.

Not sure how much of the is the hash checks vs the timestamps. We also could keep a running max timestamp rather than having to recheck all the objects again to compare for the final check on the final output freshness.

I could work on those optimizations.

I think this is the part of the previous setup that was not correct that I complained about.

Reply:
> We still need to run the backend as native code can change.  
The native code is shipped in jars + optionally a single directory for the current project. So if the native code changes, then the classpath have needed to change as well. In such case we would always trigger a new build, and the content of the jars on the classpath would override the previous dirs in our workdir. The only exception from that would involved modification of included headers on the path, and the changed libraries in e.g. `/usr/lib`. These two cases seem to be very unlikely, and even now they're out of our control. We cannot detect their modification and force a fresh linking. 
Based on that it looks to me like the backend could be safely skipped if there are no changes on the classpath and the config has not changed. However, maybe I might be missing something here.

Reply:
The project could have many sub projects with native code and jars as well just like the Scala Native project. If you are checking every jar / directory for changes then the build could be short circuited - shouldn't that be done upfront in the build not in the NIR linking? We could even skip the validation and directory creation as long as the final artifact still exists (no deletion of the build dir).

Also, then how do we know which classpath elements changed so we can re-extract native code downstream and avoid doing the hash that is being done now? Plus there is the jar/dir detection code that needs to be run. Maybe it is better to keep track of new/update/removed to just make downstream detection easier?

Reply:
Thanks @WojciechMazur @ekrich 

[[So, the initial step would be to establish a "config + classpath based" cache as we initially described in https://github.com/scala-native/scala-native/issues/3282 (If cache hit + there exists the previously generated native binary, skip the whole process and return the path to previous native binary)]], correct?

I'm thinking we can detect the change by recursively checking the mtime of the `config.classPath`, but ...
> We currently hash the jar and directories and store the result but there is some work involved to check the hash. We only unpack the code and copy to the build area the first time or if the hash differs.

do we already have a place that hashes the contents of `config.classPath`? @ekrich I'm not sure how much the timestamp is faster, but I can compare 👍 

The next step would be to implement a "NIR cache", similar to the Scala.js approach (While, I still need to fully understand this idea, need  to review the Scala.js implementation).

> Also, then how do we know which classpath elements changed so we can re-extract native code downstream and avoid doing the hash that is being done now? Plus there is the jar/dir detection code that needs to be run. Maybe it is better to keep track of new/updated/removed to just make downstream detection easier?

This should also be studied in the next step?

---

> We still need to run the backend as native code can change.
>
>  If you are checking every jar / directory for changes then the build could be short circuited - shouldn't that be done upfront in the build not in the NIR linking?
>
> The native code is shipped in jars + optionally a single directory for the current project. So if the native code changes, then the classpath have needed to change as well.

I agree with @WojciechMazur  that if we check for changes in the classpath, we should be able to detect all native code changes (by "native code", we mean NIR rather than native binary, correct?).

Native code input changes when the upstream source (sub-projects or third-party libraries) changes. In such cases, the build system should detect the changes and the scala-native front-end should re-run. This will also trigger a re-linking (NIR -> binary).

As @WojciechMazur mentioned, we cannot detect changes in C libraries located in `LD_LIBRARY_PATH` or similar locations, and users have to manually force a re-run of the linking. However, we may compromise considering that shouldn't be the case so often.

---

> We currently hash the jar and directories and store the result but there is some work involved to check the hash. We only unpack the code and copy to the build area the first time or if the hash differs.

Reply:
Skipping the whole build process if the input configuration and class paths mtime has unchanged (and output remains the same as the previous build)